### ISSUE TITLE
What is DevDiv Bugs #94657?

### ISSUE THREAD

**Issue by deesklo** [dotnet/corefx#29791]
It is referenced from [`System.Linq.Expressions.Compiler.ILGen.EmitConvertToType` method](https://github.com/dotnet/corefx/blob/v2.1-rc1/src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs#L676). That `if` branch is executed only in tests added in [this commit](https://github.com/dotnet/corefx/commit/aa808b734f14a3803843bc6164678f7569f94f90), which allowed to create expressions converting arrays when `IsAssignableFrom` returns false. And if one of array types is assignable from the other in `ILGen.EmitConvertToType`, [the previous branch](
https://github.com/dotnet/corefx/blob/v2.1-rc1/src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs#L667) in `ILGen.EmitConvertToType` gets executed.

What bug was fixed by adding the DevDiv branch, and was the branch dead code in .NET Core 2.0 and before?

**Comment by danmosemsft** [comment]
@JonHanna may have context if he added the test. I can't find that bug in the bug database I know of named DevDiv Bugs, and the reference was added before 2010, which is the oldest sources I can find right now.

**Comment by morganbr** [comment]
The .NET Framework source is at https://referencesource.microsoft.com/#System.Core/Microsoft/Scripting/Compiler/ILGen.cs,de4f63414ebe659f . I don't see any significant differences. I also didn't find any changes in this part of the file in the current source control system, so whatever this is, it's quite old.

**Comment by jkotas** [comment]
> I can't find that bug in the bug database I know of named DevDiv Bugs

This bug was in PS DevDiv Bugs database.

Here are the interesting details from the bug description:

Repro steps:

1. Compile and run the following code:

```csharp
using System;
using System.Linq;
using System.Linq.Expressions;

public class BaseClass
{
}

public class Test
{
    public static void Main()
    {
        Expression<Func<object[]>> expr = () => (object[])new BaseClass[1];
        expr.Compile(); 
    }
}
```

2. When the above code is compiled and run:

Expected: The code compiles and runs successfully

Actual: The following runtime exception is observed:

```
Unhandled Exception: System.ArgumentException: Unhandled convert: System.Object[
]
   at System.Linq.Expressions.ExpressionCompiler.GenerateNumericConversion(ILGen
erator gen, Type typeFrom, Type typeTo, Boolean isChecked)
   at System.Linq.Expressions.ExpressionCompiler.GenerateConvertToType(ILGenerat
or gen, Type typeFrom, Type typeTo, Boolean isChecked)
   at System.Linq.Expressions.ExpressionCompiler.GenerateConvert(ILGenerator gen
, UnaryExpression u)
   at System.Linq.Expressions.ExpressionCompiler.Generate(ILGenerator gen, Expre
ssion node, StackType ask)
   at System.Linq.Expressions.ExpressionCompiler.GenerateLambda(LambdaExpression
 lambda)
   at System.Linq.Expressions.ExpressionCompiler.CompileDynamicLambda(LambdaExpr
ession lambda)
   at System.Linq.Expressions.Expression`1.Compile()
   at Test.Main()
```

Description of the fix:

```
Change 2548890 on 2007/05/01 18:54:56

        This is a fix for DevDiv Bugs #94657.

        The expression tree compiler forgot to handle code generation for a
        reference conversion from one array type to another.  We now generate a
        castclass expression always for array-to-array conversions.  (We could
        be clever and optimize away the cast when going from a derived[] to a
        base[], which is legal to do without a cast.  However, this is
        dynamically generated throwaway code; the time taken to compute whether
        we can do the optimization is probably larger than the time saved!)
```

**Comment by jkotas** [comment]
It would be nice to replace the reference to the bug number with a more self-describing comment.
